xml/src/main/java/org/ehcache/xml/ConfigurationParser.java

```diff
@@ -119,30 +120,70 @@ static String replaceProperties(String originalValue, final Properties propertie
     return resolvedValue.equals(originalValue) ? null : resolvedValue;
   }
 
+  private static <T> T retrieveChild(T val1, T val2) {
+    if (val1.getClass().isInstance(val2)) {
+      return val2;
+    } else {
+      return val1;
+    }
+  }
+
   public ConfigurationParser() throws IOException, SAXException, JAXBException, ParserConfigurationException {
     Collection<Source> schemaSources = new ArrayList<>();
     schemaSources.add(new StreamSource(CORE_SCHEMA_URL.openStream()));
 
-    Set<CacheManagerServiceConfigurationParser<?>> xmlParsers = new HashSet<>();
+    Map<Class<?>, CacheManagerServiceConfigurationParser<?>> xmlParserMap = new HashMap<>();
     for (CacheManagerServiceConfigurationParser<?> parser : ClassLoading.libraryServiceLoaderFor(CacheManagerServiceConfigurationParser.class)) {
+      xmlParserMap.compute(parser.getServiceType(), (k, parserVal) -> {
+        if (parserVal == null) {
```

Comment:
This logic is repeated in 3 places. You might want to create a reusable lambda for that.